sql/expression/function/logarithm.go

```diff
@@ -0,0 +1,313 @@
+	return &Log10{expression.UnaryExpression{Child: e}}
+}
+
+func (l *Log10) String() string {
+	return fmt.Sprintf("log10(%s)", l.Child)
+}
+
+// TransformUp implements the Expression interface.
+func (l *Log10) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := l.Child.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	return f(NewLog10(child))
+}
+
+// Type returns the resultant type of the function.
+func (l *Log10) Type() sql.Type {
+	return sql.Float64
+}
+
+// IsNullable implements the sql.Expression interface.
+func (l *Log10) IsNullable() bool {
+	return l.Child.IsNullable()
+}
+
+// Eval implements the Expression interface.
+func (l *Log10) Eval(
+	ctx *sql.Context,
+	row sql.Row,
+) (interface{}, error) {
+	v, err := l.Child.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if v == nil {
+		return nil, nil
+	}
+
+	val, err := sql.Float64.Convert(v)
+	if err != nil {
+		return nil, sql.ErrInvalidType.New(reflect.TypeOf(v))
+	}
+	return computeLog(val.(float64), float64(10))
+}
+
+// Log is a function that returns the natural logarithm of a value.
+type Log struct {
+	expression.BinaryExpression
+}
+
+// NewLn creates a new Log expression.
+func NewLog(args ...sql.Expression) (sql.Expression, error) {
+	argLen := len(args)
+	if argLen == 0 || argLen > 2 {
+		return nil, sql.ErrInvalidArgumentNumber.New("1 or 2", argLen)
+	}
+
+	var right sql.Expression
+	if len(args) == 2 {
+		right = args[1]
+	}
+
+	return &Log{expression.BinaryExpression{Left: args[0], Right: right}}, nil
+}
+
+func (l *Log) String() string {
+	if l.Right == nil {
+		return fmt.Sprintf("log(%s)", l.Left)
+	}
+
+	return fmt.Sprintf("log(%s, %s)", l.Left, l.Right)
+}
+
+// TransformUp implements the Expression interface.
+func (l *Log) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	var args = make([]sql.Expression, 2)
+	arg, err := l.Left.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	args[0] = arg
+
+	args[1] = nil
+	if l.Right != nil {
+		arg, err := l.Right.TransformUp(f)
+		if err != nil {
+			return nil, err
+		}
+		args[1] = arg
+	}
+
+	expr, err := NewLog(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return f(expr)
+}
+
+// Children implements the Expression interface.
+func (l *Log) Children() []sql.Expression {
+	if l.Right == nil {
+		return []sql.Expression{l.Left}
+	}
+
+	return []sql.Expression{l.Left, l.Right}
+}
+
+// Type returns the resultant type of the function.
+func (l *Log) Type() sql.Type {
+	return sql.Float64
+}
+
+// IsNullable implements the Expression interface.
+func (l *Log) IsNullable() bool { return l.Left.IsNullable() }
+
+// Resolved implements the Expression interface.
+func (l *Log) Resolved() bool {
+	return l.Left.Resolved() && (l.Right == nil || l.Right.Resolved())
+}
+
+// Eval implements the Expression interface.
+func (l *Log) Eval(
+	ctx *sql.Context,
+	row sql.Row,
+) (interface{}, error) {
+	left, err := l.Left.Eval(ctx, row)
```

Comment:
This common block which repeats multiple times (eval + convert) we can extract to some private function.